### PR TITLE
Alistair/4705 populate user last login from auth0 events

### PR DESCRIPTION
**Pull request by alistairclark**
A few notes

* Not 100% sure if it's safe to assume that the `user_name` field in the auth0 event can be assumed to always refer to the email of the user, but I believe it can
* Worth double checking that the event time I'm using for the last login is correct.
* I've assumed that the only type of auth0 event we consider a successful login is "s"

### CODE REVIEW

File: template.yaml [hunk 1]
```diff
@@ -1135,6 +1135,60 @@ Resources:
...
+    Properties:
+      FunctionName: !Sub ${AWS::StackName}-UserRecordLastLogin
+      Description: !Sub
+        - Stack ${StackTagName} Environment ${EnvironmentTagName} Function ${ResourceName}
+        - ResourceName: UserRecordLastLogin
+      Handler: user.user_login_handler
+      Policies:
+        - AWSXrayWriteOnlyAccess
+        - DynamoDBCrudPolicy:
+            TableName: !Ref Auth0Events
```
Comment (afs25) [inline template.yaml:1149]: This will fail because there is no Auth0Events table defined in this template.

File: template.yaml [hunk 2]
```diff
@@ -1135,6 +1135,60 @@ Resources:
...
+                    - s
+                    - sce
+                    - scp
+                    - scpr
+                    - slo
+                    - ss
+                    - sv
+                    - svr
+                    - unlkdu
+                    - w
```
Comment (afs25) [inline template.yaml:1188]: This list should have only "s". We don't need to invoke the lambda for any other types.

File: template.yaml [hunk 3]
```diff
@@ -1135,6 +1135,60 @@ Resources:
...
+            EventBusName: !Ref EnvConfigeventbridgeauth0eventbusAsString
+          Metadata:
+            StackeryName: UserRecordLastLoginRule
```
Comment (afs25) [inline template.yaml:1191]: This is not needed (because we don't use Stackery anymore).

File: template.yaml [hunk 4]
```diff
@@ -1135,6 +1135,60 @@ Resources:
...
+                    - sce
+                    - scp
+                    - scpr
+                    - slo
+                    - ss
+                    - sv
+                    - svr
+                    - unlkdu
+                    - w
+            EventBusName: !Ref EnvConfigeventbridgeauth0eventbusAsString
```
Comment (afs25) [inline template.yaml:1189]: This needs to be added to the Parameters section of the template

**Review by afs25 — CHANGES_REQUESTED** [review]
Good job. See the comments for requested changes.

File: .gitignore [hunk 5]
```diff
@@ -30,6 +30,7 @@ api/local/dev_config.py
 .pytest_cache/
 tests/.pytest_cache/
 node_modules/
+.vscode/
```
Comment (this-andy) [inline .gitignore:33]: This method of exclusion from GitHub looks different from how we have done this in other repos.  Not a blocker, but let's discuss at team meeting and agree best approach.

File: api/endpoints/user.py [hunk 6]
```diff
@@ -515,3 +519,38 @@ def create_user_api(event, context):
...
+    # on successful update save audit record
+    entity_update.save()
+
+
+@utils.lambda_wrapper
+@utils.api_error_handler
+@pg_utils.db_connection_handler
+def user_login_handler(event, _):
+    detail_type = event["detail-type"]
+    assert detail_type == "Auth0 log", f"Unexpected detail-type: {detail_type}"
```
Comment (this-andy) [inline api/endpoints/user.py:543]: Not sure that we should be using assert in this way in production code.  I'd prefer to see check within try/except.
Again, let's discuss use of assert in meeting tomorrow

File: api/tests/test_data/user_data_PSFU.csv [hunk 7]
```diff
@@ -1,7 +1,7 @@
-d1070e81-557e-40eb-a7ba-b951ddb7ebdc,2018-08-17 12:10:56.798192+00,2018-08-17 12:10:56.833885+00,altha@email.co.uk,Mrs,Altha,Alcorn,,,GB,
-851f7b34-f76c-49de-a382-7e4089b744e2,2018-08-17 12:10:56.578633+00,2018-11-02 11:03:32.575029+00,bernie@email.co.uk,Mr,Bernard,Brains,,,FR,
-8518c7ed-1df4-45e9-8dc4-d49b57ae0663,2018-09-19 17:04:31.536049+00,2018-11-02 11:04:01.834929+00,clive@email.co.uk,Mr,Clive,Cresswell,1234abcd,new,IT,
-35224bd5-f8a8-41f6-8502-f96e12d6ddde,2018-08-17 12:10:56.70011+00,2018-11-02 11:06:40.450124+00,delia@email.co.uk,Ms,Delia,Davies,,,US,
-1cbe9aad-b29f-46b5-920e-b4c496d42515,2018-08-17 12:10:56.884543+00,2018-11-02 11:07:33.785406+00,eddie@email.co.uk,Mr,Eddie,Eagleton,,,GB,
-dceac123-03a7-4e29-ab5a-739e347b374d,2018-08-17 12:10:46.884543+00,2018-11-02 11:08:33.785406+00,fred@email.co.uk,Mr,Fred,Flintstone,,,US,123456
-e067ed7b-bc98-454f-9c5e-573e2da5705c,2020-08-17 12:10:46.884543+00,2020-11-02 11:08:33.785406+00,glenda@email.co.uk,Ms,Glenda,Gupta,,,IN,
\ No newline at end of file
+d1070e81-557e-40eb-a7ba-b951ddb7ebdc,2018-08-17 12:10:56.798192+00,2018-08-17 12:10:56.833885+00,altha@email.co.uk,Mrs,Altha,Alcorn,,,GB,,2018-08-17 12:10:56.833885+00
```
Comment (this-andy) [inline api/tests/test_data/user_data_PSFU.csv:1]: Would be good to have at least one of these with a null value, as this will happen in real world.

**Review by this-andy — CHANGES_REQUESTED** [review]
Would also like to get Sophie and/or Andre to comment on this PR as they have worked in these area more recently than me

File: api/tests/test_scripts/unit_tests/test_user.py [hunk 8]
```diff
@@ -627,6 +636,33 @@ def test_21_users_demo_flag_ok(self):
...
+            stack_name="thiscovery-core",
+            aws_processing_delay=10,
+            event_bus_name="auth0-event-bus"
+        )
+        if tests_running_on_aws():
+            self.assertEqual(
+                HTTPStatus.OK, result["statusCode"]
+            )
+        else:
+            self.assertEqual(HTTPStatus.NO_CONTENT, result["statusCode"])
```
Comment (afs25) [inline api/tests/test_scripts/unit_tests/test_user.py:660]: This doesn't make sense. The expected result should be the same on AWS or locally.

**Comment by alistairclark** [comment]
There are still tests which fail in this repo, but none appear to be related to my changes. 

With `TEST_ON_AWS = True` I get

```
======================================================================
FAIL: test_user_task_url_update_ok (test_user_task.TestUserTaskRedirectToInterview)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/alistair/python-projects/thiscovery-core/api/tests/test_scripts/unit_tests/test_user_task.py", line 634, in test_user_task_url_update_ok
    self.assertEqual(
AssertionError: 'https://test-afs25-video.thiscovery.org?[44 chars]6j29' != None
```

With `TEST_ON_AWS = False` `test_project_status_for_user_default.test_user_c_project_status` fails because the real/expected URLs in a dictionary are not equal.

**Review by afs25 — APPROVED** [review]

**Review by this-andy — APPROVED** [review]

**Pull request merged**